### PR TITLE
refactor: nightfall-node requested changes

### PR DESCRIPTION
**Pull request by imagobea**
## What does this implement/fix? Explain your changes.

Refactor from transaction creation to submission:

- Call a function to create tx (includes gas estimation), using mutex

- Submit transaction and add the resulting promise to a queue (endpoint will return the `tx hash`, which is available right after signing)

- Actions that need to happen after the transaction has settled, will only take place if the tx was successful

## Does this close any currently open issues?

#1260 

## What commands can I run to test the change? 

Like the other `nightfall-node` PRs, you can rely on curl to make calls to the optimist, or run the tests. Note that now routes are protected with an API key (instead of the sha256 derived from the proposer key).

## Any other comments?

Inspired by #1093

### CODE REVIEW

File: nightfall-optimist/src/services/transaction-sign-send.mjs [hunk 1]
```diff
@@ -0,0 +1,46 @@
+import { web3 } from '@polygon-nightfall/common-files/utils/contract.mjs';
+import { estimateGas, estimateGasPrice } from '@polygon-nightfall/common-files/utils/gas.mjs';
+import logger from '@polygon-nightfall/common-files/utils/logger.mjs';
+import app from '../app.mjs';
```
Comment (imagobea) [inline nightfall-optimist/src/services/transaction-sign-send.mjs:4]: Had to try this anyways :joy:

File: cli/lib/nf3.mjs [hunk 2]
```diff
@@ -1066,6 +1066,7 @@ class Nf3 {
       fee,
     });
     logger.debug(`Proposer /update response ${res}`);
+    return res?.data;
```
Comment (imagobea) [inline cli/lib/nf3.mjs:1069]: :memo:

Have updated tests with call to `/register`, so that these "await" for the register tx to be mined (event `NewCurrentProposer` can only be guaranteed in an env with 1 proposer).

Does `nf3` need to be updated? Tests using `nf3.registerProposer` seem to be passing anyway.
Keep an eye.

File: test/utils.mjs [hunk 3]
```diff
@@ -522,3 +522,9 @@ export const emptyL2 = async ({ nf3User, web3, logs }) => {
   }
   await new Promise(resolve => setTimeout(resolve, 6000));
 };
+
+export async function isTransactionMined(txHash, web3) {
```
Comment (israelboudoux) [inline test/utils.mjs:526]: The name of this function is a bit confusing because it gives the impression that a boolean value is gonna be returned. Maybe changing it to `waitTransactionToBeMined` or something similar will match with the function's goal.

File: test/utils.mjs [hunk 4]
```diff
@@ -522,3 +522,9 @@ export const emptyL2 = async ({ nf3User, web3, logs }) => {
...
+  const receipt = await web3.eth.getTransactionReceipt(txHash);
+  if (receipt !== null) return;
+  await isTransactionMined(txHash, web3);
```
Comment (israelboudoux) [inline test/utils.mjs:529]: This function is recursive, but there isn't a limit for retrying, maybe one could add a timeout, even if it is a small one, just to avoid too many recursive calls and consequently prevent the heap from being out of memory.

File: nightfall-optimist/src/services/transaction-sign-send.mjs [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+export async function getAddressNonce(ethAddress) {
+  let nonce;
+  try {
+    nonce = await web3.eth.getTransactionCount(ethAddress);
+  } catch (err) {
+    logger.error({
+      msg: 'Error obtaining address nonce',
+      err,
+    });
```
Comment (israelboudoux) [inline nightfall-optimist/src/services/transaction-sign-send.mjs:29]: If an error happens, the `nonce` will be `null`. At this point, you could throw the error you captured.

File: nightfall-optimist/src/services/transaction-sign-send.mjs [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+const { GAS, GAS_MULTIPLIER, GAS_ESTIMATE_ENDPOINT, GAS_PRICE, GAS_PRICE_MULTIPLIER } = config;
+
+const nonceMutex = new Mutex();
+
+const isListening = async () => {
+  try {
+    await web3.eth.net.isListening();
+  } catch (error) {
+    throw new Error('Web3 ws is disconnected, try again later');
```
Comment (israelboudoux) [inline nightfall-optimist/src/services/transaction-sign-send.mjs:17]: We could get the message from the `error` variable and pass in the exception.

File: nightfall-optimist/src/services/block-assembler.mjs [hunk 7]
```diff
@@ -168,25 +169,51 @@ export async function conditionalMakeBlock(proposer) {
...
+        );
+
+        // Submit tx and update db if tx is successful
+        // CHECK - db ops
+        txsQueue.push(async () => {
+          try {
+            const receipt = await sendSignedTransaction(signedTx);
+            logger.debug({ msg: 'Block proposed', receipt });
+
+            await removeTransactionsFromMemPool(block.transactionHashes);
```
Comment (israelboudoux) [inline nightfall-optimist/src/services/block-assembler.mjs:198]: Maybe we could make use of `Promise.all([])` taking the these three calls inside it. e.g. 
```
await Promise.all([
  removeTransactionsFromMemPool(block.transactionHashes),
  removeCommitmentsFromMemPool(transactions.map(t => t.commitments.filter(c => c !== ZERO)).flat(Infinity)),
  removeNullifiersFromMemPool(transactions.map(t => t.nullifiers.filter(c => c !== ZERO)).flat(Infinity)),
]);
```
The code would be a bit cleaner and the behaviour it is the same.

File: nightfall-optimist/src/app.mjs [hunk 8]
```diff
@@ -17,18 +16,21 @@ import {
...
 
 const app = express();
 
 const ethPrivateKey = environment.PROPOSER_KEY;
 const { address } = web3.eth.accounts.privateKeyToAccount(ethPrivateKey);
+const nonce = await getAddressNonce(address);
 
 app.set('ethPrivateKey', ethPrivateKey);
 app.set('ethAddress', address);
+app.set('nonce', nonce);
```
Comment (israelboudoux) [inline nightfall-optimist/src/app.mjs:31]: Sorry, I didn't get the idea around this constant `nonce`.

File: nightfall-optimist/src/routes/proposer.mjs [hunk 9]
```diff
@@ -91,20 +95,35 @@ router.post('/register', auth, async (req, res, next) => {
...
+        ethPrivateKey,
+        ethAddress,
+        proposersContractAddress,
+        txDataToSign,
+        stake,
+      );
+
+      // Submit tx and update db if tx is successful
+      // CHECK - I think the code beyond L132 can be removed (?) then db op could be moved here
+      txsQueue.push(async () => {
```
Comment (israelboudoux) [inline nightfall-optimist/src/routes/proposer.mjs:116]: This piece of code is repeated many times, changing only the log message sometimes and the calls inside. Maybe we can think in something to reuse it and make it less verbose. ;)

**Review by israelboudoux — CHANGES_REQUESTED** [review]
LGTM, there are just a few things to check.

File: common-files/utils/gas.mjs [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+ *
+ * The underlying RPC `eth_estimateGas` will check the balance of the sender,
+ * this means that even though the call doesn't consume any gas,
+ * the `from` address must have enough gas to execute the tx
+ *
+ * @async
+ * @function estimateGas
+ * @param {object} tx A tx object
+ * @param {Web3} web3
+ * @param {number} gasDefault Has default
```
Comment (velkoskis) [inline common-files/utils/gas.mjs:17]: @param {number} gasDefault default gas price

File: nightfall-optimist/src/app.mjs (hunk 8 above)
Comment (velkoskis) [reply nightfall-optimist/src/app.mjs:31]: I think that it is needed to create a signed transaction for making a block if the current proposer is the actual running proposer.

File: nightfall-optimist/src/services/transaction-sign-send.mjs [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+  // Check if web3 ws is available
+  await isListening();
+
+  logger.debug({ msg: 'Create transaction object...' });
+
+  let signedTx;
+  await nonceMutex.runExclusive(async () => {
+    // Update nonce if necessary
+    const _nonce = await getAddressNonce(from);
+    if (nonce < _nonce) {
```
Comment (velkoskis) [inline nightfall-optimist/src/services/transaction-sign-send.mjs:45]: why is this needed?

**Review by velkoskis — APPROVED** [review]
Good job on the implementation. LGTM

File: test/utils.mjs (hunk 3 above)
Comment (imagobea) [reply test/utils.mjs:526]: True. Somehow I thought that given the recursive nature, the name might be okay. But I agree that using `wait` is better. Thank you.

File: test/utils.mjs (hunk 4 above)
Comment (imagobea) [reply test/utils.mjs:529]: Agree. But not entirely sure about what to do.
When mentioning `setTimeOut`, do you mean inside the recursion? ie

```
// await getTransactionReceipt
// if condition for exiting
// else setTimeOut(waitTransactionToBeMined, 3000)
```

A counter might be possible too. Please let me know :pray:

File: nightfall-optimist/src/services/transaction-sign-send.mjs (hunk 6 above)
Comment (imagobea) [reply nightfall-optimist/src/services/transaction-sign-send.mjs:17]: Tbh I think this implementation is a bit off, I'll take another look :pray:

File: nightfall-optimist/src/app.mjs (hunk 8 above)
Comment (imagobea) [reply nightfall-optimist/src/app.mjs:31]: I guess @velkoskis that is what I was trying to do, but since I am not refreshing the value stored in the state of the app, this doesn't make much sense. Thank @israelboudoux , I will re-think this one (alongside nonce error-handling above).
Thanks again both

File: nightfall-optimist/src/routes/proposer.mjs (hunk 9 above)
Comment (imagobea) [reply nightfall-optimist/src/routes/proposer.mjs:116]: :eyes: That is true, but I would appreciate if I could do this in another PR :pray:

File: nightfall-optimist/src/services/transaction-sign-send.mjs (hunk 11 above)
Comment (daveroga) [reply nightfall-optimist/src/services/transaction-sign-send.mjs:45]: I think this is to synchronize the nonce if another transaction has been sent before processing this transaction. If not then a revert will be thrown because the account has a higher nonce.

File: test/utils.mjs (hunk 4 above)
Comment (daveroga) [reply test/utils.mjs:529]: When I say a timeout I refer to a timeout from the perspective of the caller of the function `waitTransactionToBeMined` in in the case of the loop function it would be a counter.

File: nightfall-optimist/src/routes/proposer.mjs (hunk 9 above)
Comment (israelboudoux) [reply nightfall-optimist/src/routes/proposer.mjs:116]: Sure! :D

File: nightfall-optimist/src/services/transaction-sign-send.mjs (hunk 11 above)
Comment (velkoskis) [reply nightfall-optimist/src/services/transaction-sign-send.mjs:45]: @daveroga isn't that what nonceMutex.runExclusive() is taking care of?

File: test/utils.mjs (hunk 4 above)
Comment (imagobea) [reply test/utils.mjs]: Okay: removed recursion, added timer and counter.

File: nightfall-optimist/src/services/transaction-sign-send.mjs (hunk 6 above)
Comment (imagobea) [reply nightfall-optimist/src/services/transaction-sign-send.mjs:17]: FYI `isListening` can return a boolean but also fail for many reasons
For the happy path, I am now returning the result from calling the web3 method, then I only proceed with the transaction if the result is "truthy"
If there's an error, the exception will be raised re-using the error message as per your suggestion

**Comment by imagobea** [comment]
This is it sirs, thank you for your comments @velkoskis @daveroga @israelboudoux (and apologies again, like I said the page went "stale" and I was replying to some comments while missing others).
Have requested a re-review from Isra, but feel free to comment or ask further questions.
Thanks again :pray:

File: nightfall-optimist/src/routes/proposer.mjs [hunk 12]
```diff
@@ -134,7 +156,13 @@ router.post('/register', auth, async (req, res, next) => {
         }
       }
     }
-    res.json({ receipt });
+
+    if (signedTx) {
```
Comment (israelboudoux) [inline nightfall-optimist/src/routes/proposer.mjs:160]: This `if` is not needed. If `signedTx` is `null`, then something wrong happened, right?

File: nightfall-optimist/src/services/block-assembler.mjs [hunk 13]
```diff
@@ -221,13 +251,14 @@ export async function conditionalMakeBlock(proposer) {
...
         // blocks with them
-        await removeTransactionsFromMemPool(block.transactionHashes);
-        await removeCommitmentsFromMemPool(
-          transactions.map(t => t.commitments.filter(c => c !== ZERO)).flat(Infinity),
-        );
-        await removeNullifiersFromMemPool(
-          transactions.map(t => t.nullifiers.filter(c => c !== ZERO)).flat(Infinity),
-        );
+
+        // await removeTransactionsFromMemPool(block.transactionHashes);
```
Comment (israelboudoux) [inline nightfall-optimist/src/services/block-assembler.mjs:255]: Pls, remove these code commented out.

File: nightfall-optimist/src/services/transaction-sign-send.mjs [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+    const isListening = await web3.eth.net.isListening();
+    return isListening;
+  } catch (error) {
+    throw new Error(error.message);
+  }
+};
+
+export async function getAddressNonce(ethAddress) {
+  try {
+    const nonce = await web3.eth.getTransactionCount(ethAddress);
```
Comment (israelboudoux) [inline nightfall-optimist/src/services/transaction-sign-send.mjs:24]: You can return directly.

File: nightfall-optimist/src/services/transaction-sign-send.mjs [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+import logger from '@polygon-nightfall/common-files/utils/logger.mjs';
+import { Mutex } from 'async-mutex';
+
+const { GAS, GAS_MULTIPLIER, GAS_ESTIMATE_ENDPOINT, GAS_PRICE, GAS_PRICE_MULTIPLIER } = config;
+
+const nonceMutex = new Mutex();
+
+const isWeb3Listening = async () => {
+  try {
+    const isListening = await web3.eth.net.isListening();
```
Comment (israelboudoux) [inline nightfall-optimist/src/services/transaction-sign-send.mjs:15]: You can return directly.

File: nightfall-optimist/src/services/transaction-sign-send.mjs [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+  } catch (error) {
+    throw new Error(error.message);
+  }
+}
+
+// TODO document
+export async function createSignedTransaction(nonce, ethPrivateKey, from, to, data, value = 0) {
+  // Check if web3 ws is available
+  const result = await isWeb3Listening();
+  if (result) {
```
Comment (israelboudoux) [inline nightfall-optimist/src/services/transaction-sign-send.mjs:35]: I'd invert this `if` so that we return the negative asap since it is just a piece of code, while the code inside the `if` statement is pretty big. e.g. 
`if(! result) {
    throw new Error('Web3 ws not listening, try again later');
}
... // the rest...

File: nightfall-optimist/src/app.mjs (hunk 8 above)
Comment (imagobea) [reply nightfall-optimist/src/app.mjs:31]: We finally agreed to skip all nonce tracking for now and monitor behaviour

File: nightfall-optimist/src/routes/proposer.mjs (hunk 12 above)
Comment (imagobea) [reply nightfall-optimist/src/routes/proposer.mjs]: Addressed in these commits [d8c542f](https://github.com/EYBlockchain/nightfall_3/pull/1262/commits/d8c542f8a22ab5163b9921975cfb2eb1203646f5), [5fb74e3](https://github.com/EYBlockchain/nightfall_3/pull/1262/commits/5fb74e347080f688f49bb85abe3e9f9c953d3d9e) (I cleaned up all nonce stuff as discussed)

File: nightfall-optimist/src/services/block-assembler.mjs (hunk 13 above)
Comment (imagobea) [reply nightfall-optimist/src/services/block-assembler.mjs:249]: Will address in #1198

File: nightfall-optimist/src/services/transaction-sign-send.mjs (hunk 16 above)
Comment (imagobea) [reply nightfall-optimist/src/services/transaction-sign-send.mjs:26]: Of course

File: nightfall-optimist/src/services/transaction-sign-send.mjs (hunk 15 above)
Comment (imagobea) [reply nightfall-optimist/src/services/transaction-sign-send.mjs:15]: Like we said it is not actually possible to return straight from an `asyn` call within a `try/catch` block (see https://jakearchibald.com/2017/await-vs-return-vs-return-await/).

In any case - like you pointed out - both methods `isWeb3Listening` and `getAddressNonce` were used locally and I've ended up removing both.

Thanks

File: nightfall-optimist/src/services/transaction-sign-send.mjs (hunk 14 above)
Comment (imagobea) [reply nightfall-optimist/src/services/transaction-sign-send.mjs]: Removed, see comment re `isWeb3Listening`

File: nightfall-optimist/src/services/transaction-sign-send.mjs (hunk 15 above)
Comment (israelboudoux) [reply nightfall-optimist/src/services/transaction-sign-send.mjs]: Amazing, thank you!

**Review by israelboudoux — APPROVED** [review]
LGTM
